### PR TITLE
Use type hints and ThreadPoolExecutor instead of bare Thread

### PR DESCRIPTION
**Pull request by rabinadk1**
1. Use type hints for various functions
2. Use `ThreadPoolExecutor` to make threads automatically join the calling thread
3. Use it to join hanging threads
4. Write a README with instructions on how to run the code.
5. Move common methods to `commons.py`
6. Add `.gitignore` and Code Scanning for GitHub
7. Formatted the code using Black

### CODE REVIEW

**Comment by nrasadi** [comment]
Great effort! Thanks :clap:
I will review your PR, and merge it asap.

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+name: "CodeQL"
+
+on:
+  push:
+    branches: [ main ]
```
Comment (rabinadk1) [inline .github/workflows/codeql-analysis.yml:5]: You need to change this and the one below on line 8 to your default branch, which is `master` for now.

**Comment by nrasadi** [comment]
Apologies for the late response! I have been really busy lately.

Again, your modifications look beautiful. I love the idea of using typing hints. Thanks.
However, something's buggy. When I increase the number of branches, things don't go correctly, but It works on the current version.
Could you please check it?

Also, my code has a bug in printing messages in the inspector side. (It doesn't show those transfers that their `send message` arrives later than its corresponding  `send message`.
I fix this issue in the upcoming commit. bug gets fixes by modifying the `find_transfer_message` and `get_messages` methods in the `Inspector` class. Please consider this in your future changes.

Bests

**Comment by rabinadk1** [comment]
Superseded by #2

**Pull request closed without merging**